aiida_quantumespresso/workflows/pw/relax.py

```diff
@@ -143,13 +157,56 @@ def setup(self):
         self.ctx.is_converged = False
         self.ctx.iteration = 0
 
+        self.ctx.relax_inputs = AttributeDict(self.exposed_inputs(PwBaseWorkChain, namespace='base'))
+        self.ctx.relax_inputs.pw.parameters = self.ctx.relax_inputs.pw.parameters.get_dict()
+
+        self.ctx.relax_inputs.pw.parameters.setdefault('CONTROL', {})
+        self.ctx.relax_inputs.pw.parameters['CONTROL']['restart_mode'] = 'from_scratch'
+
+        # Adjust the inputs for the chosen relaxation scheme
+        if 'relaxation_scheme' in self.inputs:
+            if self.inputs.relaxation_scheme.value in ('relax', 'vc-relax'):
+                self.ctx.relax_inputs.pw.parameters['CONTROL']['calculation'] = self.inputs.relaxation_scheme.value
+            else:
+                raise ValueError('unsupported value for the `relaxation_scheme` input.')
+
+        # Set the meta_convergence and add it to the context
+        self.ctx.meta_convergence = self.inputs.meta_convergence.value
+        volume_cannot_change = (
+            self.ctx.relax_inputs.pw.parameters['CONTROL']['calculation'] in ('scf', 'relax') or
+            self.ctx.relax_inputs.pw.parameters['CELL']['cell_dofree'] == 'shape'
```

Comment:
I think this has still the capacity to fail now, no? I don't think it is guaranteed that `['CONTROL']['calculation']` or `['CELL']['cell_dofree']` exists. For the former, we should probably add a validator in the spec because the user now *has* to indicate the calculation mode (before we did it through the `relax_type` input). The second one is sometimes not defined and that is ok, so we should do something like `self.ctx.relax_inputs.pw.parameters.get('CELL', {}).get('cell_dofree', None) == 'shape'`

Reply:
Right, of course. 🤦 

One issue with the validator is that the user might still use the `relaxation_scheme` input, and I don't think I can check if this input is provided in the validator of the `base` input? So the user might not provide the `'calculation'`, in the `base` input, but the `relaxation_scheme` instead.

Reply:
you can put the validator on the top-level namespace for now with a comment that once `relaxation_scheme` is removed, it can be moved to the `parameters` port. In the top-level you will have access to all inputs

Reply:
I've added the validator which checks if the `base` input has `['pw']['parameters']['CONTROL']['calculation']` specified. It might be a little confusing to a user that still relies on `relaxation_scheme` and doesn't specify the `calculation` input, but I maybe this is ok? Or is there an elegant solution to this?

Reply:
yeah, as my previous comment said, you can move the validator to the top level inputs which has access to all
```
def validate_inputs(inputs, ctx):
    """Validator for top level namespace."""
        parameters = inputs['base']['pw']['parameters'].get_dict()

        if 'relaxation_scheme' not in inputs and 'calculation' not in parameters.get('CONTROL', {}):
            return 'The parameters in `base.pw.parameters` do not specify the required key `CONTROL.calculation`.'

def define(cls, spec):
    spec.input()
    ....
    spec.inputs.validator = validate_inputs
```

Note also that you should not raise an exception from a validator but simply return a string with the error message

Reply:
Ah, it seems Github hadn't updated the comments when I replied to yours, sorry about that!